File: vuu/src/main/scala/org/finos/vuu/benchmark/tree/TreeBenchmark.scala
package org.finos.vuu.benchmark.tree

import org.finos.toolbox.jmx.MetricsProviderImpl
import org.finos.toolbox.lifecycle.LifecycleContainer
import org.finos.toolbox.time.{Clock, DefaultClock}
import org.finos.vuu.benchmark.BenchmarkHelper.{createBigTable, createTreeBuilder}
import org.finos.vuu.core.table._
import org.finos.vuu.provider.JoinTableProviderImpl
import org.finos.vuu.viewport.tree.{Tree, TreeBuilder}
import org.openjdk.jmh.annotations._
import org.openjdk.jmh.runner.Runner
import org.openjdk.jmh.runner.options.OptionsBuilder

import java.io.IOException
import java.util.concurrent.TimeUnit

@State(Scope.Benchmark)
class TreeBenchmark {

  var table: SimpleDataTable = null
  var treeBuilder: TreeBuilder = null
  implicit val clock: Clock = new DefaultClock
  implicit val lifecycle: LifecycleContainer = new LifecycleContainer
  implicit val metrics: MetricsProviderImpl = new MetricsProviderImpl
  val joinProvider = JoinTableProviderImpl() // new EsperJoinTableProviderImpl()

  val tableContainer = new TableContainer(joinProvider)
  @Setup(Level.Invocation)
  def setup(size: Int): Unit = {
    table = createBigTable(size)
    treeBuilder = createTreeBuilder(table)
  }

  def doTree(table: SimpleDataTable, treeBuilder: TreeBuilder): Tree = {
    treeBuilder.buildEntireTree()
  }

  @Benchmark
  @OutputTimeUnit(TimeUnit.MILLISECONDS)
  @Warmup(iterations = 5)
  @Measurement(iterations = 10)
  @BenchmarkMode(Array(Mode.AverageTime))
  @throws[IOException]
  def treeLargeTable(): Unit = {
    doTree(table, treeBuilder)
  }

  def main(args: Array[String]): Unit = {
    val opts = new OptionsBuilder()
      .include(classOf[TreeBenchmark].getSimpleName)
      .warmupIterations(5)
      .measurementIterations(5)
      //.forks(1)
      .build
    new Runner(opts).run
  }

}

object TreeRun {
  def main(args: Array[String]): Unit = {
    val benchmark = new TreeBenchmark()
    benchmark.setup(1_000_000)
    benchmark.treeLargeTable()
  }
}

File: benchmark/src/main/java/org/finos/vuu/benchmark/TreeBenchmark.java
package org.finos.vuu.benchmark;

import org.openjdk.jmh.annotations.*;

import java.io.IOException;
import java.util.concurrent.TimeUnit;

@State(Scope.Benchmark)
public class TreeBenchmark {

  public static void main(String[] args) throws Exception {
    final TreeBenchmark bm = new TreeBenchmark();
    bm.tableSize = 10000;
    bm.setup();
    bm.treeLargeTable();
  }

  private org.finos.vuu.benchmark.tree.TreeBenchmark benchmark;

  @Param({ "10000", "100000", "500000", "1000000", "2000000", "5000000" })
  public int tableSize;

  @Setup
  public void setup(){
    benchmark = new org.finos.vuu.benchmark.tree.TreeBenchmark();
    System.out.println("Tree Benchmark - Setup: tableSize=" + tableSize);
    benchmark.setup(tableSize);
  }

    @Benchmark
    @OutputTimeUnit(TimeUnit.MILLISECONDS)
    @Warmup(iterations = 1)
    @Measurement(iterations = 1)
    @Fork(1)
    @BenchmarkMode(Mode.SampleTime)
    public void treeLargeTable() throws IOException {
      System.out.println("TreeBenchmark java - treeLargeTable");
      benchmark.treeLargeTable();
    }
}


